src/content/en/fundamentals/engage-and-retain/push-notifications/sending-messages.markdown

```diff
@@ -3,93 +3,343 @@ layout: shared/narrow
+1. At some later point, your app server sends a message to the messaging server.
+
+## Generating the applicationServerKey
+
+There are several things you need to know about the `applicationServerKey`:
+
+* It's the public key portion of a public/private key pair generated on your application server.
+* The key pair should be usable with an elliptic curve digital signature (ECDSA) over the P-256 curve.
+* Your app should pass key to the server as an array of eight-bit unsigned integers.
+* It's required by a spec called Voluntary Application Server Identification for Web Push (VAPID), which we'll discuss in the [sending messages](sending-messages) section. 
+
+You can find an example for generating this in the [web-push node library](https://github.com/web-push-libs/web-push/). It looks something like this:
+
+{% highlight javascript %}
+function generateVAPIDKeys() {
+  var curve = crypto.createECDH('prime256v1');
+  curve.generateKeys();
+
+  return {
+    publicKey: curve.getPublicKey(),
+    privateKey: curve.getPrivateKey(),
+  };
+}
+{% endhighlight %}
+
+## Anatomy of a subscription object
+
+Earlier we said that a subscription object must be stringified and passed to the 
+server, but we didn't tell you what was in the subscription object. That's 
+because the client doesn't do anything with it. The server does.  
+  
+The subscription object looks like this:  
+
+{% highlight javascript %}
+{  
+  "endpoint": "https://example.com/push-service/send/dbDqU8xX10w:APA91b...",  
+  "keys": {  
+    "auth": "qLAYRzG9TnUwbprns6H2Ew==",  
+    "p256dh": 
+"BILXd-c1-zuEQYXH\\_tc3qmLq52cggfqqTr\\_ZclwqYl6A7-RX2J0NG3icsw..."  
+  }  
 }
 {% endhighlight %}
 
-What's in this?
+What's in this?  
+  
+**endpoint**—Contains two parts: the URL of the messaging service used by the
+subscribing browser followed by unique identifier for the user. This is called a
+subscription ID or  a registration ID. This tells your server how to identify
+you the messaging  server.
+  
+**keys**—Encryption keys used for encrypting data passed to the service worker 
+messages. It contains the following:
+
+* **auth**-A 16 byte authentication secret generated by the browser.
+* **p256dh**-A 65 byte containing the vapidPublicKey created on your app server.
+
+**Note:** In many of the relevant specs bytes are called octets. The term is 
+used because of legacy and communication systems where bytes may not always be 8 
+bits long.
+
+## Creating the message
+
+This is where things start to get a little crazy. In this section we're no 
+longer in your client app. We're on the app server where we are going to create 
+and send a message to the client. There's much to keep track of. 
+
+Before continuing, let's review what we have and where it came from.
+
+* **Subscription object** - his came from the client. It contains the endpoint 
+  of the messaging server, a copy of the public key (`applicationServerKey`), and an authentication 
```

Comment:
You don't have the applicationServerKey in the subscription object. You have a public key used to encrypt the payload.